# Módulo Extra

## Aula 1

In [45]:
import pandas as pd
import numpy as np

In [2]:
df_cadastros = pd.read_csv('https://raw.githubusercontent.com/alura-cursos/Avaliacao_Credito/main/dados/clientes_cadastrados.csv')
df_cadastros.head()

ID_Cliente Genero  Tem_Carro  Tem_Casa_Propria  Qtd_Filhos  \
0     5008804      H          1                 1           0   
1     5008805      H          1                 1           0   
2     5008806      H          1                 1           0   
3     5008808      M          0                 1           0   
4     5008809      M          0                 1           0   

   Rendimento_Anual   Categoria_de_renda Grau_Escolaridade   Estado_Civil  \
0          427500.0            Empregado   Ensino superior  Uniao estavel   
1          427500.0            Empregado   Ensino superior  Uniao estavel   
2          112500.0            Empregado      Ensino medio         Casado   
3          270000.0  Associado comercial      Ensino medio       Solteiro   
4          270000.0  Associado comercial      Ensino medio       Solteiro   

        Moradia  Idade  Tem_Celular  Tem_telefone_trabalho  Tem_telefone_fixo  \
0       Alugado   33.0            1                      1                  0   
1       Alugado   33.0            1                      1                  0   
2  Casa propria   59.0            1                      0                  0   
3  Casa propria   52.0            1                      0                  1   
4  Casa propria   52.0            1                      0                  1   

   Tem_email   Ocupacao  Tamanho_Familia  Anos_empregado  
0          0     Outros              2.0            12.4  
1          0     Outros              2.0            12.4  
2          0  Seguranca              2.0             3.1  
3          1      Venda              1.0             8.4  
4          1      Venda              1.0             8.4

In [5]:
df_cadastros['ID_Cliente'].value_counts()

7045794    2
7636389    2
7046068    2
7052812    2
7742298    2
          ..
5709384    1
5727815    1
5729862    1
6780485    1
6293503    1
Name: ID_Cliente, Length: 438510, dtype: int64

In [9]:
df_cadastros[df_cadastros['ID_Cliente'].duplicated(keep=False)].head()

ID_Cliente Genero  Tem_Carro  Tem_Casa_Propria  Qtd_Filhos  \
421349     7602432      M          0                 0           0   
422660     7052783      H          1                 1           2   
423702     7838075      M          1                 1           0   
424480     7053557      H          1                 1           1   
425306     7023651      M          0                 0           0   

        Rendimento_Anual   Categoria_de_renda           Grau_Escolaridade  \
421349          117000.0          Pensionista             Ensino superior   
422660          166500.0            Empregado                Ensino medio   
423702          315000.0  Associado comercial             Ensino superior   
424480          193500.0            Empregado                Ensino medio   
425306          225000.0  Associado comercial  Ensino superior incompleto   

       Estado_Civil       Moradia  Idade  Tem_Celular  Tem_telefone_trabalho  \
421349       Casado  Casa propria   68.0            1                      0   
422660       Casado  Casa propria   44.0            1                      1   
423702     Solteiro  Casa propria   29.0            1                      0   
424480       Casado  Casa propria   29.0            1                      1   
425306     Solteiro  Casa propria   28.0            1                      0   

        Tem_telefone_fixo  Tem_email          Ocupacao  Tamanho_Familia  \
421349                  0          0            Outros              2.0   
422660                  0          1          Gerencia              4.0   
423702                  0          1            Outros              1.0   
424480                  0          0  Equipe principal              3.0   
425306                  0          0     Contabilidade              1.0   

        Anos_empregado  
421349         -1000.7  
422660             7.4  
423702             4.5  
424480             6.2  
425306             3.3

In [15]:
id_repetidos = df_cadastros[df_cadastros['ID_Cliente'].duplicated(keep=False)]['ID_Cliente']
id_repetidos.head()

421211    7702516
421268    7602432
421349    7602432
421464    7836971
421698    7213374
Name: ID_Cliente, dtype: int64

In [26]:
df_cadastros_ID_unicos = df_cadastros.drop(id_repetidos.index)
df_cadastros_ID_unicos.head()

ID_Cliente Genero  Tem_Carro  Tem_Casa_Propria  Qtd_Filhos  \
0     5008804      H          1                 1           0   
1     5008805      H          1                 1           0   
2     5008806      H          1                 1           0   
3     5008808      M          0                 1           0   
4     5008809      M          0                 1           0   

   Rendimento_Anual   Categoria_de_renda Grau_Escolaridade   Estado_Civil  \
0          427500.0            Empregado   Ensino superior  Uniao estavel   
1          427500.0            Empregado   Ensino superior  Uniao estavel   
2          112500.0            Empregado      Ensino medio         Casado   
3          270000.0  Associado comercial      Ensino medio       Solteiro   
4          270000.0  Associado comercial      Ensino medio       Solteiro   

        Moradia  Idade  Tem_Celular  Tem_telefone_trabalho  Tem_telefone_fixo  \
0       Alugado   33.0            1                      1                  0   
1       Alugado   33.0            1                      1                  0   
2  Casa propria   59.0            1                      0                  0   
3  Casa propria   52.0            1                      0                  1   
4  Casa propria   52.0            1                      0                  1   

   Tem_email   Ocupacao  Tamanho_Familia  Anos_empregado  
0          0     Outros              2.0            12.4  
1          0     Outros              2.0            12.4  
2          0  Seguranca              2.0             3.1  
3          1      Venda              1.0             8.4  
4          1      Venda              1.0             8.4

In [17]:
df_cadatros_ID_unicos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 438463 entries, 0 to 438556
Data columns (total 18 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   ID_Cliente             438463 non-null  int64  
 1   Genero                 438463 non-null  object 
 2   Tem_Carro              438463 non-null  int64  
 3   Tem_Casa_Propria       438463 non-null  int64  
 4   Qtd_Filhos             438463 non-null  int64  
 5   Rendimento_Anual       438463 non-null  float64
 6   Categoria_de_renda     438463 non-null  object 
 7   Grau_Escolaridade      438463 non-null  object 
 8   Estado_Civil           438463 non-null  object 
 9   Moradia                438463 non-null  object 
 10  Idade                  438463 non-null  float64
 11  Tem_Celular            438463 non-null  int64  
 12  Tem_telefone_trabalho  438463 non-null  int64  
 13  Tem_telefone_fixo      438463 non-null  int64  
 14  Tem_email              438463 non-nu

### Desafio 1: Explorar IDs repetidos (Atualização de cadastro? Erro?).

In [19]:
df_cadastros.iloc[id_repetidos.index].sort_values('ID_Cliente')

ID_Cliente Genero  Tem_Carro  Tem_Casa_Propria  Qtd_Filhos  \
426818     7022197      H          1                 1           3   
425023     7022197      M          0                 1           0   
431545     7022327      M          0                 1           0   
431911     7022327      H          1                 1           0   
425486     7023108      H          1                 1           1   
...            ...    ...        ...               ...         ...   
426563     7836711      M          0                 1           2   
421464     7836971      H          1                 0           1   
428620     7836971      M          0                 1           0   
422068     7838075      H          0                 1           0   
423702     7838075      M          1                 1           0   

        Rendimento_Anual   Categoria_de_renda Grau_Escolaridade  \
426818          135000.0            Empregado      Ensino medio   
425023          450000.0  Associado comercial   Ensino superior   
431545          135000.0  Associado comercial      Ensino medio   
431911          256500.0  Associado comercial   Ensino superior   
425486           67500.0            Empregado      Ensino medio   
...                  ...                  ...               ...   
426563          292500.0            Empregado   Ensino superior   
421464          157500.0            Empregado      Ensino medio   
428620          103500.0            Empregado      Ensino medio   
422068          337500.0  Associado comercial      Ensino medio   
423702          315000.0  Associado comercial   Ensino superior   

         Estado_Civil       Moradia  Idade  Tem_Celular  \
426818         Casado  Casa propria   33.0            1   
425023       Separado  Casa propria   54.0            1   
431545       Solteiro  Casa propria   40.0            1   
431911         Casado  Casa propria   59.0            1   
425486         Casado  Casa propria   42.0            1   
...               ...           ...    ...          ...   
426563         Casado  Casa propria   38.0            1   
421464         Casado  Casa propria   38.0            1   
428620  Uniao estavel  Casa propria   37.0            1   
422068         Casado  Casa propria   50.0            1   
423702       Solteiro  Casa propria   29.0            1   

        Tem_telefone_trabalho  Tem_telefone_fixo  Tem_email          Ocupacao  \
426818                      0                  0          1  Construcao Civil   
425023                      0                  0          1            Outros   
431545                      0                  0          0   Alta tecnologia   
431911                      0                  0          1  Equipe principal   
425486                      1                  0          0  Equipe principal   
...                       ...                ...        ...               ...   
426563                      0                  1          0     Contabilidade   
421464                      0                  0          0            Outros   
428620                      0                  1          0             Venda   
422068                      0                  0          1         Motorista   
423702                      0                  0          1            Outros   

        Tamanho_Familia  Anos_empregado  
426818              5.0             2.0  
425023              1.0             4.9  
431545              1.0            14.5  
431911              2.0             4.6  
425486              3.0             4.6  
...                 ...             ...  
426563              4.0            12.0  
421464              3.0            15.1  
428620              2.0             7.7  
422068              2.0             3.5  
423702              1.0             4.5  

[94 rows x 18 columns]

Observando os dados onde os ID são repetidos, parece que relmente foi um erro, onde a maior parte das features são diferentes. Como por exemplo o ID 7022197, onde uma das linhas se refere a um homem de 33 anos, e a segunda a uma mulher de 54 anos. Fato que torna difícil acreditar que foi alguma atualização, e sim algum erro do sistema. Por esta razão o melhor a se fazer nesse caso deve ser eliminar esses dados mesmo.

### Desafio 2: Explorar as colunas (features) identificar continuas, categóricas e binárias.

In [28]:
df_cadastros_ID_unicos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 438463 entries, 0 to 438556
Data columns (total 18 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   ID_Cliente             438463 non-null  int64  
 1   Genero                 438463 non-null  object 
 2   Tem_Carro              438463 non-null  int64  
 3   Tem_Casa_Propria       438463 non-null  int64  
 4   Qtd_Filhos             438463 non-null  int64  
 5   Rendimento_Anual       438463 non-null  float64
 6   Categoria_de_renda     438463 non-null  object 
 7   Grau_Escolaridade      438463 non-null  object 
 8   Estado_Civil           438463 non-null  object 
 9   Moradia                438463 non-null  object 
 10  Idade                  438463 non-null  float64
 11  Tem_Celular            438463 non-null  int64  
 12  Tem_telefone_trabalho  438463 non-null  int64  
 13  Tem_telefone_fixo      438463 non-null  int64  
 14  Tem_email              438463 non-nu

In [35]:
features_int = df_cadastros_ID_unicos.select_dtypes('int64').columns
features_int

Index(['ID_Cliente', 'Tem_Carro', 'Tem_Casa_Propria', 'Qtd_Filhos',
       'Tem_Celular', 'Tem_telefone_trabalho', 'Tem_telefone_fixo',
       'Tem_email'],
      dtype='object')

In [57]:
for feature in features_int:
    print(feature + ': Valores Únicos' + str(df_cadastros_ID_unicos[feature].unique())+ '\n')

ID_Cliente: Valores Únicos[5008804 5008805 5008806 ... 6841878 6842765 6842885]

Tem_Carro: Valores Únicos[1 0]

Tem_Casa_Propria: Valores Únicos[1 0]

Qtd_Filhos: Valores Únicos[ 0  1  3  2  4  5 14 19  7  9 12  6]

Tem_Celular: Valores Únicos[1]

Tem_telefone_trabalho: Valores Únicos[1 0]

Tem_telefone_fixo: Valores Únicos[0 1]

Tem_email: Valores Únicos[0 1]



Observando as variáveis classificadas como inteiro percebemos que a maioria é binária. Porém temos o ID_cliente onde cada um se refere a um cliente. A quantidade de filhos que está variando entre 0 e 19, e além disso, temos a variável tem celular, onde, para todos os clientes, o valor é igual a 1, ou seja, todos os clientes possuem celular, e essa variável aparenta ser desnecessária para o nosso problema.

In [39]:
features_cont = df_cadastros_ID_unicos.select_dtypes('float64').columns
features_cont

Index(['Rendimento_Anual', 'Idade', 'Tamanho_Familia', 'Anos_empregado'], dtype='object')

In [56]:
for feature in features_cont:
    print(feature + ': mínimo e máximo  ' + str(df_cadastros_ID_unicos[feature].min()) + ' - ' + str(df_cadastros_ID_unicos[feature].max())+ '\n')

Rendimento_Anual: mínimo e máximo  26100.0 - 6750000.0

Idade: mínimo e máximo  21.0 - 69.0

Tamanho_Familia: mínimo e máximo  1.0 - 20.0

Anos_empregado: mínimo e máximo  -1000.7 - 48.0



As variávies classificadas como números contínuos parecem ter sido classificadas corretamente, onde todas representam um valor numérico.

In [53]:
features_cat = df_cadastros_ID_unicos.select_dtypes('object').columns
features_cat

Index(['Genero', 'Categoria_de_renda', 'Grau_Escolaridade', 'Estado_Civil',
       'Moradia', 'Ocupacao'],
      dtype='object')

In [55]:
for feature in features_cat:
    print(feature + ': Valores únicos  ' + str(df_cadastros_ID_unicos[feature].unique()) + '\n')

Genero: Valores únicos  ['H' 'M']

Categoria_de_renda: Valores únicos  ['Empregado' 'Associado comercial' 'Pensionista' 'Servidor publico'
 'Estudante']

Grau_Escolaridade: Valores únicos  ['Ensino superior' 'Ensino medio' 'Ensino superior incompleto'
 'Ensino fundamental' 'Pos graduacao']

Estado_Civil: Valores únicos  ['Uniao estavel' 'Casado' 'Solteiro' 'Separado' 'Viuvo']

Moradia: Valores únicos  ['Alugado' 'Casa propria' 'Habitacao publica' 'Mora com os pais'
 'Cooperativa habitacional' 'Apartamento comercial']

Ocupacao: Valores únicos  ['Outros' 'Seguranca' 'Venda' 'Contabilidade' 'Construcao Civil'
 'Gerencia' 'Motorista' 'Equipe principal' 'Alta tecnologia' 'Limpeza'
 'Servico privado' 'Cozinha' 'Baixa qualificacao' 'Medicina'
 'Secretariado' 'Garcom' 'RH' 'Corretor imobiliario' 'TI']



Todas as variávies classificadas como categóricas parecem ter sido classificadas corretamente também.